File: tests/fixtures/csharp-to-cpp-port/sample-app/Models/Device.cs
using System;
using SampleApp.Util;

namespace SampleApp.Models
{
    public class Device
    {
        public int Id { get; }
        public string Name { get; set; }
        public DeviceKind Kind { get; set; }
        public int? LastSeenTick { get; set; }

        public Device(int id, string name, DeviceKind kind)
        {
            if (name == null) throw new ArgumentNullException(nameof(name));
            Id = id;
            Name = name;
            Kind = kind;
            LastSeenTick = null;
        }

        public bool IsOnline => LastSeenTick.HasValue;

        public string Describe()
        {
            string seen = LastSeenTick.HasValue ? LastSeenTick.Value.ToString() : "never";
            return $"[{StringHelpers.PadId(Id)}] {Name} ({Kind}) seen={seen}";
        }

        public override string ToString()
        {
            return Describe();
        }
    }
}

File: tests/fixtures/csharp-to-cpp-port/expected-cpp/Models/Device.cpp
// Ported from Models/Device.cs
#include "Models/Device.h"

#include "PortSupport.h"
#include "Util/StringHelpers.h"

namespace SampleApp::Models {

Device::Device(int id, std::wstring name, DeviceKind kind)
    : id_(id), name_(std::move(name)), kind_(kind), lastSeenTick_(std::nullopt)
{
    // C# threw ArgumentNullException when name == null; std::wstring cannot be null, so no check.
}

std::wstring Device::Describe() const
{
    const std::wstring seen = lastSeenTick_.has_value()
        ? PortSupport::ToWString(*lastSeenTick_)
        : std::wstring(L"never");
    return L"[" + Util::StringHelpers::PadId(id_) + L"] " + name_
        + L" (" + DeviceKindToString(kind_) + L") seen=" + seen;
}

std::wstring Device::ToString() const
{
    return Describe();
}

} // namespace SampleApp::Models
